src/metabase/query_processor/util/add_alias_info.clj

```diff
@@ -235,32 +235,38 @@
             (m/find-first (fn [[_field an-id-or-name _opts]]
                             (= an-id-or-name id-or-name))
                           field-exports)))
-        ;; look for a matching expression clause with the same name if still no match
+        ;; otherwise if this is a nominal field literal ref then look for matches based on the string name used
         (when-let [field-name (let [[_ id-or-name] field-clause]
                                 (when (string? id-or-name)
                                   id-or-name))]
-          (or ;; Expressions by exact name.
+          (or ;; First, look for Expressions or fields from the source query stage whose `::desired-alias` matches the
+              ;; name we're searching for.
+              (m/find-first (fn [[tag _id-or-name {::keys [desired-alias], :as _opts} :as _ref]]
+                              (when (#{:expression :field} tag)
+                                (= desired-alias field-name)))
+                            all-exports)
```

Comment:
This is the actual fix, it should have looked at both Field and expression refs for a matching `::desired-alias`, not just expression refs.